### DISCUSSION TITLE
Where to report missing dependency on formula?

### DISCUSSION THREAD

**Discussion by sparr** [Homebrew/discussions#4939] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.22
ORIGIN: https://github.com/Homebrew/brew
HEAD: 251a098fbf1c02ca95f53b4806e5b6da8d5e4a80
Last commit: 5 days ago
Core tap JSON: 30 Nov 19:46 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nano
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit dunno
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

After installing `cmake` I encountered multiple failures with the included bash completions, including errors during shell loading. I have reported those upstream, in hopes of getting them fixed there. However, in the mean time, the `cmake` formula has a dependency on `bash-completion` to provide some helper functions for its bash completions.
I've read the instructions on the Issue creation pages for Homebrew/brew and Homebrew/homebrew-core and can't figure out the right category/template to use to report this, so I'm here looking for guidance on where and how to report this problem.

**Comment by osalbahr** [comment]
[Here](https://github.com/Homebrew/homebrew-core/blob/master/Formula/c/cmake.rb#L30) is where dependencies are listed, are you interested in [making a PR](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request)?

**Comment by sparr** [comment]
Adding the dependency could have undesired side effects. The comments in that file say 

> `# The completions were removed because of problems with system bash`

so maybe they actually should be removed? Or that comment should be removed and the dependency added. Or the script refactored upstream to not have that dependency (https://gitlab.kitware.com/cmake/cmake/-/issues/25461). Or... some other solution that might come up in an Issue about this situation, if I could figure out where to put it.

**Reply by osalbahr** [reply]
Ah, good catch. Yeah, not sure which way to go about it. You might want to try to find the PR where the comment was added for relevant discussion.

To better understand the issue, could you provide error logs and steps to reproduce it on my end?

**Reply by sparr** [reply]
1. `brew remove bash-completion` (if you have it installed, and you can reinstall it afterward)
2. `brew install cmake`
3. `bash`
4. `cmake ` followed by the tab key to trigger completions

```
bash: _split_longopt: command not found
bash: _filedir: command not found
```

**Reply by osalbahr** [reply]
I couldn't replicate the error:
```console
% bash
bash-5.2$ cd $(mktemp -d)
bash-5.2$ touch test
bash-5.2$ cmake test # hitting <tab> completed "test"
```

Perhaps it might be system-dependent?
```console
bash-5.2$ brew config
HOMEBREW_VERSION: 4.1.23
ORIGIN: https://github.com/Homebrew/brew
HEAD: fc259153591a309e2b6fee6f971eb61b16a3b6c0
Last commit: 13 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: e123e970a39ad5274f0e56fec5d8ccc3877983fb
Core tap last commit: 5 days ago
Core tap branch: master
Core tap JSON: 05 Dec 08:13 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.M710wUHjiS/org.xquartz:0
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.43.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.6.1-x86_64
CLT: 15.1.0.0.1.1699597245
Xcode: N/A
```

**Reply by sparr** [reply]
@osalbahr Can you try without creating the file? Also can you confirm you don't have bash-completion installed? Also what is your output from `command -V _split_longopt`? If it's found, can you tell where it's coming from?

**Reply by osalbahr** [reply]
1. Without the file, tab completion does nothing (except the bell sound Terminal makes)
2. I do not have bash-completion installed
3. 
```console
% brew list | grep bash
bash
% command -V _split_longopt
_split_longopt not found
```

**Reply by sparr** [reply]
How about in a directory with many varied files? I think I was just running the commands in my home directory, which has subdirectories, hidden files, symlinks, regular files, ...

**Reply by sparr** [reply]
Ahh, Per @cho-m's comments below, I suspect your shell isn't `source`ing the cmake completion script.

**Comment by cho-m** [comment]
> However, in the mean time, the `cmake` formula has a dependency on `bash-completion` to provide some helper functions for its bash completions.

That would make it difficult for macOS users who want to use latest `bash + bash-completion@2` as the completions formulae conflict.

---

One option is to move completions to `bash-completion@2` directory instead of v1 directory as that directory won't get loaded for users who copied config from docs: https://docs.brew.sh/Shell-Completion#configuring-completions-in-bash

It isn't unusual for completions to assume installation of some `bash-completion*` script given path completions get installed into. However, since Homebrew docs do include manually sourcing those completions, that can result in these issues. Personally, I wouldn't recommend doing the following:
```bash
    for COMPLETION in "${HOMEBREW_PREFIX}/etc/bash_completion.d/"*
    do
      [[ -r "${COMPLETION}" ]] && source "${COMPLETION}"
    done
```

and instead use the corresponding formula.

---

EDIT: https://github.com/Homebrew/homebrew-core/pull/156445

**Reply by cho-m** [reply]
Fun situation is that future release of `bash-completion@2` will no longer have `_split_longopt` - https://github.com/scop/bash-completion/commit/19a3798a0bf46ad9e39a055c4113eb5e9e9cb98b so `cmake` upstream will need to handle this.

**Reply by sparr** [reply]
Just to be clear, this is probably an improvement in general, but it doesn't address the problem where there's no source of completion helper functions at all.

**Reply by cho-m** [reply]
That's up to the upstream whether they want to support non-`bash-completion` users or not. Not something that can be fix discussing it here.

**Reply by sparr** [reply]
It can be fixed by adding bash-completion as a dependency of cmake, here.

**Reply by cho-m** [reply]
That is an unnecessary dependency. Also, as I previously mentioned, it will cause breakage for all users who installed newer `bash-completion@2` due to conflict.

Thus, it is not an acceptable option.

**Reply by sparr** [reply]
I'm new to brew packaging. Does `bash-completion@2` not satisfy a dependency on `bash-completion[with some suffix meaning "any version"]` that could be added here?

**Reply by cho-m** [reply]
No. `depends_on "<formula>"` is for a specific formula. If we specify `depends_on "bash-completion"`, then that means v1 will be installed on user systems and users who installed `bash-completion@2` will get an error message like:
```
==> Installing cmake dependency: bash-completion
Error: Cannot install bash-completion because conflicting formulae are installed.
  bash-completion@2: because each are different versions of the same formula

Please `brew unlink bash-completion@2` before continuing.
```

**Reply by sparr** [reply]
Reading some issues and feature requests around implementing multiple/optional/fallback dependencies, it seems like the priority here is more on guaranteeing that a single particular combination of packages works, rather than ever possible viable combination. It also seems like causing conflicts is seen as an acceptable option in some scenarios similar to this one.

**Reply by cho-m** [reply]
Again, it is not acceptable to add a conflict here. We try to avoid conflicts as much as possible and `cmake` is an essential build package that we should not impact everyone who wants to use a newer `bash`.

It does not make any sense to add `bash-completion` dependency.

If you want a priority, it will be removing completion. That was what the original behavior was before I added completions and I was planning to backtrack the commit but never got around to it

See: https://github.com/Homebrew/homebrew-core/pull/108228
And: https://github.com/Homebrew/homebrew-core/pull/113569/files#r1035371043
